partner/WHMCS/modules/servers/jetpack/jetpack.php

```diff
@@ -0,0 +1,288 @@
+        logModuleCall('jetpack', __FUNCTION__, $params, $e->getMessage(), $e->getTraceAsString());
+        return $e->getMessage();
+    }
+}
+
+/**
+ * Equivalent to partner/cancel. Cancel a Jetpack plan using
+ * using a Jetpack Hosting partner account.
+ *
+ * @param array $params
+ * @return string
+ * @throws Exception
+ */
+function jetpack_TerminateAccount(array $params)
+{
+    $validation_status = validate_required_fields($params);
+    if ($validation_status !== true) {
+        return $validation_status;
+    }
+
+    try {
+        $access_token = get_access_token($params);
+        $stripped_url = preg_replace("(^https?://)", "", $params['customfields']['Site URL']);
+        $clean_url = str_replace('/', '::', $stripped_url);
+
+        $request_url = 'https://public-api.wordpress.com/rest/v1.3/jpphp/' . $clean_url . '/partner-cancel';
+        $response = make_api_request($request_url, $access_token);
+        if ($response->success === true) {
+            return 'success';
+        } elseif ($response->success === false) {
+            return 'JETPACK MODULE: Unable to terminate this Jetpack plan as it has likely already been cancelled';
+        }
+    } catch (Exception $e) {
+        logModuleCall('jetpack', __FUNCTION__, $params, $e->getMessage(), $e->getTraceAsString());
+        return $e->getMessage();
+    }
+}
+
+/**
+ * Get a Jetpack partner access token using the client_id
+ * and client secret stored when the product was created in
+ * the WHMCS product settings.
+ *
+ *
+ * @param $params
+ * @return mixed
+ * @throws Exception
+ */
+function get_access_token($params)
+{
+
+    $oauthURL = "https://public-api.wordpress.com/oauth2/token";
+
+    $credentials = array (
+        'client_id' => $params['configoption1'],
+        'client_secret' => $params['configoption2'],
+        'grant_type' => 'client_credentials',
+        'scope' => 'jetpack-partner'
+    );
+
+    $response = make_api_request($oauthURL, null, $credentials);
+    if (isset($response->access_token)) {
+        return $response->access_token;
+    } else {
+        return $response;
+    }
+}
+
+
+/**
+ * Make an API request for authenticating and provisioning or
+ * cancelling a Jetpack plan
+ *
+ * @param $url
+ * @param $data
+ * @param string $auth
+ * @return mixed
+ * @throws Exception
+ */
+function make_api_request($url, $auth = null, $data = null)
+{
+    if (isset($auth)) {
+        $auth = "Authorization: Bearer " . $auth;
+    }
+
+    $curl = curl_init();
+    curl_setopt_array($curl, array(
+        CURLOPT_HTTPHEADER => array($auth),
+        CURLOPT_URL => $url,
+        CURLOPT_RETURNTRANSFER => true,
+        CURLOPT_ENCODING => "",
+        CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
+        CURLOPT_POSTFIELDS => $data,
+        CURLOPT_CUSTOMREQUEST => "POST"
+    ));
+
+    $response = curl_exec($curl);
+    $http_code = curl_getinfo($curl, CURLINFO_HTTP_CODE);
+
+    if ($response && $http_code >= 400) {
```

Comment:
What if the response is 301 or 100?

Reply:
@nb great question! I understood what a successful response should be but handled edge cases inadequately including these 2 that you mentioned. This is no longer in the make_api_request function and will be handled by the calling function so there's more room for error management and clarity on what to do when specific errors occur that were legitimate responses and not curl errors.